src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java

```diff
@@ -43,13 +49,19 @@
 public final class ZipkinHttpClientImpl implements HttpClient {
 
     private final HttpClient delegate;
+    private final CurrentTraceContext currentTraceContext;
+    private final ThreadLocalSpan span;
+    private final BiFunction<WrappedRequestSpec, TraceContext, Span> nextSpan;
     private final HttpClientHandler<WrappedRequestSpec, Integer> handler;
     private final TraceContext.Injector<MutableHeaders> injector;
 
     @Inject
     public ZipkinHttpClientImpl(final HttpClient delegate, final HttpTracing httpTracing) {
         this.delegate = delegate;
-        this.handler = HttpClientHandler.create(httpTracing, new HttpAdapter());
+        this.span = ThreadLocalSpan.create(httpTracing.tracing().tracer());
```

Comment:
On the usage of `ThreadLocalSpan`, looking at the Javadoc for that class, re. the callbacks:

```
Provided the library guarantees these run on the same thread, you can simply propagate the
 * result of {@link Tracer#withSpanInScope(Span)} from the starting callback to the closing one.
```
Don't know enough about how Ratpack does things to be able to verify that we have these guarantees - @llinder, do you know? I think in our case, the "onStart/onFinish" callbacks would be in the request method:

https://github.com/adriancole/ratpack-zipkin/blob/07f7e18f0f7d46dd6c83bcbca16bee74bc0694aa/src/main/java/ratpack/zipkin/internal/ZipkinHttpClientImpl.java#L98-L99

Reply:
This seems like it should work ok.  Since `ThreadLocalSpan` is deferring to the tracer and thus the configured `CurrentTraceContext`, it should still be on a Ratpack managed thread and thus the Registry values should be in a reliable state.

Then again there are a lot of "should"s in there :)  Without really digging into Ratpack code I can't say what will happen with 100% certainty though :)

The pattern of using a partial function with the ThreadLocalSpan tidies things up nicely though and is worth a try.